ssh/src/extension.ts

```diff
@@ -46,41 +53,226 @@ export function activate(context: vscode.ExtensionContext): void {
+		const jumpHost = getJumpHost();
+		let destinationOptions: ConnectOptions  = { 
+			username: options.username,
+			host: getServerHost() || options.host,
+			port: getServerPort() || options.port
+		}
 
-		logger.log(`Connecting to ${username}@${host} port ${port}`);
+		// Populate jump host falling back on server settings for user and port
+		let jumpOptions: ConnectOptions | null = null;
+		if (jumpHost) {
+			const jumpUser = getJumpUser();
+			const jumpPort = getJumpPort();
+			jumpOptions = {
+				username: jumpUser ? jumpUser : options.username,
+				host: jumpHost,
+				port: jumpPort ? jumpPort: options.port
+			}
+		}
 
-		conn.on('x11', handleX11);
+		if (preferConfig) {
+			const sshConfig = resolveHome(getSshConfig());
+			const host = getServerHost() || options.host;
+
+			logger.log(`Attempting to parse ${sshConfig} for ssh configuration of ${host}`);
+
+			try {
+				const data = fs.readFileSync(resolveHome(getSshConfig()))
+
+				const config = SSHConfig.parse(data.toString());
+
+				let hostConfig = null;
+				for (const line of config) {
+					if (line.param === 'Host') {
+						if (line.value === host) {
+							hostConfig = line.config;
+							break;
+						} else {
+							const hostNameMatch = line.config.find((line:any) => line.param === 'HostName' && line.value === host);
+							if (hostNameMatch) {
+								logger.log(`${host} matches ${line.value}`);
+								hostConfig = line.config;
+								break;
+							}
+						}
+					}
+				}
+
+				if (!hostConfig) {
+					throw new Error(`Could not find ${host}`);
+				}
+
+				const configUser = hostConfig.find((line:any) => line.param === 'User');
+				const configPort = hostConfig.find((line:any) => line.param === 'Port');
+				const configHost = hostConfig.find((line:any) => line.param === 'HostName');
+				const configProxy = hostConfig.find((line:any) => line.param === 'ProxyCommand');
+
+				destinationOptions = {
+					username: configUser ? configUser.value : options.username,
+					port: configPort ? configPort.value : options.port,
+					host: configHost ? configHost.value : options.host
+				}
+
+				// If it's set to proxy
+				if (configProxy) {
+					logger.log(`Attempting to parse ProxyCommand for ssh configuration of ${host}`);
+
+					const proxyCommand: string = configProxy.value;
+					// For now we just handle ssh forwards.  I don't know what other edge cases are out there
+					if (proxyCommand.startsWith('ssh')) {
+						const args = proxyCommand.split(' ');
+						const jumpHostArg = args[args.length-1];
+
+						// forward channel -W
+						const forwardArg = args.indexOf('-W');
+						if (forwardArg > 0) { 
+							logger.log(`Attempting to parse forwarding for ssh configuration of ${host}`);
+
+							const hostAndPort = args[forwardArg + 1].split(':');
+							// %h is the config host, %p is the config port
+							destinationOptions.host = hostAndPort[0] === '%h' ? destinationOptions.host : hostAndPort[0];
```

Comment:
Would be nice to include a check like
```ts
const hostAndPortStr = args[forwardArg + 1];
const hostAndPort = hostAndPortStr.split(':');
if (hostAndPort.length < 2) {
  throw new Error(`Invalid destination "${hostAndPortStr}". Expected a string of the form "host:port"`);
}
```

Reply:
Yah, good point.  Since that is a strict argument I skipped the check, but we're the ones parsing so we'll need to be explicit.